frappe/erpnext
# feat: batch-wise item valuation

### Description
This PR switches the legacy valuation for batched items to Batch-wise Item Valuation. Legacy valuation method for
batched items did not use to consider batches.

To understand the difference, consider the following incoming transactions, assume the same item *I*:
|# |Type   |Qty|Rate|Warehouse|Batch No|
|--|-------|---|----|---------|--------|
|1 |Inward |1  |100 |A        |X       |
|2 |Inward |1  |150 |A        |Y       |

If an outward entry is to be made:
|# |Type   |Qty|Warehouse|Batch No|
|--|-------|---|---------|--------|
|3 |Outward| -1|A        |Y       |

what will the item valuation be if the item uses *FIFO Valuation*, *Moving
Average Valuation*?

<details>
<summary>
<b>Before :: FIFO: 100, MA: 125</b>
</summary>
Valuation is considered on an item and warehouse basis so after transactions 1 and
2 this is how the queues will look:

|# |*Key* | Stock Queue       |
|--|------|-------------------|
|1 |(I, A)|[(1,100)]         |
|2 |(I, A)|[(1,100), (1,150)] |

Since outward entry (*# 3*) is for item *I* and warehouse *A* the last 
queue fetched for valuation is keyed by I and A. Batch is ignored here.

*Note: A key isn't implicitly used*
</details>

<details>
<summary>
<b>After :: FIFO: 150, MA: 150</b>
</summary>
Valuation is considered on an item, batch, and warehouse basis, so after
transactions 1 and 2 this is how the queues will look:

|# |*Key*    | Stock Queue |
|--|---------|-------------|
|1 |(I, X, A)|[(1,100)]    |
|2 |(I, Y, A)|[(1,150)]    |

Since outward entry (*# 3*) is for item *I*, batch *Y*, and warehouse *A* the last 
queue fetched for valuation is keyed by I, Y, and A. Batch is not ignored here.

*Note: A key isn't implicitly used*
</details>

### Additional Notes
- *Batchwise Balance History* has been refactored to account for the changes in this PR.
- Existing batches will continue to use the legacy valuation method.
- Batchwise-item valuation will apply only to batches created after this PR has been rolled out.
- The following situation is possible (by using the `use_batchwise_valuation` flag on _Batch_ and _Stock Ledger Entry_):
  1. Item A has two batches X, Y, and Z.
  2. Y uses batches-wise valuation.
  3. X and Z use legacy valuation.
  4. The valuation of Y is completely isolated from that of X and Z.

### Steps to test (QA Team)
1. Create a batched item.
2. Make _inward_ and _outward_ transactions using the created item 
    - _Inward_: Purchase Receipts, Material Receipt Stock Entries, etc.
    - _Outward_: Material Issue, Delivery Notes, etc.
    - Vary the Batch No across transactions (check the example above).
    - Vary the Rate across the transaction.
3. Check the stock ledger entry values
    - The stock queue, valuation should not change between two transactions if the batch no doesn't match.
    - Example:
      1. _Inward_ :: batch A, rate 100, qty 1: stock queue will be `[[1, 100]]`.
      2. _Inward_ :: batch B, rate 120, qty 2: stock queue will be `[[2, 120]]` not `[[1,100], [2, 120]]`.
4. Valuation Rate is used on outward transactions, this too you will get from the stock ledger entries; transactions on batch B shouldn't affect the valuation rate of batch A.

Stuff that may help:
- https://frappe.io/kb/stock/calculation-of-valuation-rate-in-fifo-and-moving-average
- https://frappe.io/kb/stock/how-valuation-rate-is-calculated-for-an-item


'no-docs'
